rfcs/RFC-Upload-From-Dropbox.md

```diff
@@ -0,0 +1,367 @@
+1. A user can check the status of their upload from the DP Browser App.
+1. A user can restart a failed upload from the DP Browser App.
+1. The user must be authenticated before uploading.
+1. A user who has not yet agreed to the Portal policies must not be able to upload their file.
+1. Supports the upload of AnnData, Loom, or Seurat v3 RDS. See [Reference](https://github.com/chanzuckerberg/corpora-data-portal/blob/main/backend/schema/corpora_schema.md#implementations).
+
+### Nonfunctional requirements
+
+1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
+1. A user can upload a single file up to 30GB in size.
+1. Shared links referring to more than one file will be rejected.
+1. The status of an upload must be tracked and queryable.
+
+## Detailed Design | Architecture | Implementation
+
+![Upload Design](https://app.lucidchart.com/publicSegments/view/a74f747f-d071-41ea-949f-52f38a98747a/image.png)
+Figure 1: The Upload Design
+
+### Cloud to Cloud Upload Flow
+
+Cloud to cloud upload requires the user to provide a shareable link for a file on their DropBox from which DP can upload
+the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure 3.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 3:** Architecture of a user uploading from their cloud to our cloud.
+
+#### 1. Get Share Link
+
+The user must generate a sharable link from their DropBox. Sharable links should only point to a single file, not a
+directory of files.
+
+#### 2. Share Link
+
+Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
+by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
+check the status of the upload
+
+#### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by DropBox are often not direct
+links to the files and need to be parsed into the correct shape for downloading. The correct URL will be generated
+as part of the sanitization process. The backend that runs this should also be given minimal privileges to further protect from
+server-side request forgery. How to download from a shared link on [Dropbox](https://help.dropbox.com/files-folders/share/force-download).
+
+The backend will also create an entry in the upload table using the dataset_id as the key, and set the status to waiting.
+
+#### 4. Queue
+
+The queue stores the links that need to be processed. It buffers the downstream servers from download requests. This
+allows us to spin up servers as needed. It also gives us the ability to retry a download if one of the servers to fails
+upload the file.
+
+##### Why Might the Upload Fail
+
+- The downloading server ran out of storage.
```

Comment:
Can this feasibly happen given that the entirety of the dataset will never be on the server itself? It seems like we'll be streaming in data in chunks which will prevent this?

Reply:
For dropbox, this likely won't be a problem